This notebook contains a neural network that detects the musical key of a piece based on the musical pitches of the notes. All the pieces were taken from midiworld (https://www.midiworld.com/search/?q=POP).

In [1]:
import pandas as pd
import numpy as np
import tensorflow as tf
from tensorflow import keras
from sklearn.model_selection import train_test_split
from midi_to_csv import convert_data
from clean_data import clean


In [ ]:
# ensure data is properly converted and cleaned.
convert_data()
clean()

We create a neural network with multiple hidden layers and utilize keras's Adam optimzer during training. 

In [6]:
# Load the dataframe
df = pd.read_csv('clean_data/keys.csv')

# Split the dataframe into input (notes) and output (musical key) variables
X = df.iloc[:, 1:].values  # Notes (numerical representation)
y = df.iloc[:, 0].values   # Musical key

# Convert the musical key labels to integers from 0 to 23 inclusive
key_to_int = {key: i for i, key in enumerate(np.unique(y))}
y = np.array([key_to_int[key] for key in y])

# Normalize the input data
X = X / np.max(X)

# Define the neural network architecture
model = keras.Sequential([
    keras.layers.Dense(64, activation='relu', input_shape=(X.shape[1],)),
    keras.layers.Dense(64, activation='relu'),
    keras.layers.Dense(24, activation='softmax')
])

# Compile the model
model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Train the model
model.fit(X_train, y_train, epochs=15, batch_size=200, validation_data=(X_test, y_test))

# Evaluate the model
_, accuracy = model.evaluate(X_test, y_test)

print('Test accuracy:', accuracy)

Epoch 1/15
1/1 [==============================] - 0s 237ms/step - loss: nan - accuracy: 0.6933 - val_loss: nan - val_accuracy: 0.7105
Epoch 2/15
1/1 [==============================] - 0s 23ms/step - loss: nan - accuracy: 0.6933 - val_loss: nan - val_accuracy: 0.7105
Epoch 3/15
1/1 [==============================] - 0s 24ms/step - loss: nan - accuracy: 0.6933 - val_loss: nan - val_accuracy: 0.7105
Epoch 4/15
1/1 [==============================] - 0s 21ms/step - loss: nan - accuracy: 0.6933 - val_loss: nan - val_accuracy: 0.7105
Epoch 5/15
1/1 [==============================] - 0s 29ms/step - loss: nan - accuracy: 0.6933 - val_loss: nan - val_accuracy: 0.7105
Epoch 6/15
1/1 [==============================] - 0s 31ms/step - loss: nan - accuracy: 0.6933 - val_loss: nan - val_accuracy: 0.7105
Epoch 7/15
1/1 [==============================] - 0s 30ms/step - loss: nan - accuracy: 0.6933 - val_loss: nan - val_accuracy: 0.7105
Epoch 8/15
1/1 [==============================] - 0s 31ms/step - los

Overall, our model yields an accuracy of 71%. Fiddling with epoch sizes/numbers seems to yield the smae result. This is likely because pieces can often deviate from their musical key, either through modulation or embellishment. Moreover, music from different eras is likely to less/more adhere to traditional musical strcuture. Thus, more accurate models likely need to consider the style and era the music was produced during,